Update Docs with Finalize Block (#760)
cometbft/cometbft: docs/guides/go-built-in.md

go 1.20

require (

> github.com/cometbft/cometbft v0.38.0

Reviewer: might be good to preserve indentation

indentation might be due to linter/formatter (go fmt)

package main

import (

> abcitypes "github.com/cometbft/cometbft/abci/types"

Reviewer: same comment re: indentation

Note that `CheckTx` does not execute the transaction, it only verifies that the transaction could be executed. We do not know yet if the rest of the network has agreed to accept this transaction into a block.

Finally, make sure to add the `bytes` package to the `import` stanza at the top of `app.go`:

```go
import(
    "bytes"

    "github.com/dgraph-io/badger/v3"
    abcitypes "github.com/cometbft/cometbft/abci/types"
)
```

### 1.3.3 FinalizeBlock

> `FinalizeBlock` is an ABCI method introduced in CometBFT `v0.38.0`. This replaces the functionality provided previously (pre-`v0.38.0`) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.

Reviewer 2: ```suggestion
When the CometBFT consensus engine has decided on the block, the block is transferred to the
application via `FinalizeBlock`.
`FinalizeBlock` is an ABCI method introduced in CometBFT `v0.38.0`. This replaces the functionality provided previously (pre-`v0.38.0`) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.
```

Reviewer 3: I must insist that I don't see the point of mentioning the old methods here. This is an introductory guide, for people that don't know the older versions.
Upgrading from older versions should be a tutorial in itself.

Author: I see @lasarojc. But I don't really see any harm in mentioning previous methods. If a legacy user from 2019 or 2020 comes in he will get some context on what happened to old methods. New users, they might just not care and move ahead.

This method is responsible for executing the block and returning a response to the consensus engine.

> The `ProcessProposal` method provides the transactions to the state machine before the block is finalized.
> This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. 

Reviewer 2: ```suggestion
```

Providing a single `FinalizeBlock` method to signal the finalization of a block simplifies the ABCI interface and increases flexibility in the execution pipeline.

The `FinalizeBlock` method executes the block, including any necessary transaction processing and state updates, and returns a `ResponseFinalizeBlock` object which contains any necessary information about the executed block.

> **Note:** `FinalizeBlock` only prepares the update to be made and does not change the state of the application. The change of state happens in a later stage i.e. is `commit` phase.

Reviewer 2: ```suggestion
**Note:** `FinalizeBlock` only prepares the update to be made and does not change the state of the application. The change of state happens in a later stage i.e. in `commit` phase.
```

Note that we check the validity of the transaction _again_ during `FinalizeBlock`.

```go
func (app *KVStoreApplication) FinalizeBlock(_ context.Context, req *abcitypes.RequestFinalizeBlock) (*abcitypes.ResponseFinalizeBlock, error) {
    var txs = make([]*abcitypes.ExecTxResult, len(req.Txs))

    app.onGoingBlock = app.db.NewTransaction(true)
    for i, tx := range req.Txs {
        if code := app.isValid(tx); code != 0 {
            log.Printf("Error: invalid transaction index %v", i)
            txs[i] = &abcitypes.ExecTxResult{Code: code}
        } else {
            parts := bytes.SplitN(tx, []byte("="), 2)
            key, value := parts[0], parts[1]
            log.Printf("Adding key %s with value %s", key, value)

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }

            log.Printf("Successfully added key %s with value %s", key, value)

            // Marshal the ExecTxResult to the txs slice element
            txs[i] = &abcitypes.ExecTxResult{}
        }
    }

    return &abcitypes.ResponseFinalizeBlock{

>     TxResults:        txs,

Reviewer 2: Indent?

    app.onGoingBlock = app.db.NewTransaction(true)
    for i, tx := range req.Txs {
        if code := app.isValid(tx); code != 0 {
            log.Printf("Error: invalid transaction index %v", i)
            txs[i] = &abcitypes.ExecTxResult{Code: code}
        } else {
            parts := bytes.SplitN(tx, []byte("="), 2)
            key, value := parts[0], parts[1]
            log.Printf("Adding key %s with value %s", key, value)

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }

            log.Printf("Successfully added key %s with value %s", key, value)

            // Marshal the ExecTxResult to the txs slice element
            txs[i] = &abcitypes.ExecTxResult{}
        }
    }

    return &abcitypes.ResponseFinalizeBlock{
    TxResults:        txs,
    }, nil
}
```

Note that we check the validity of the transaction _again_ during `DeliverTx`.
Transactions are not guaranteed to be valid when they are delivered to an
application, even if they were valid when they were proposed.
This can happen if the application state is used to determine transaction

> validity. Application state may have changed between the initial execution of `CheckTx`

Reviewer 3: this comment is still relevant, once adapted to finalizeBlock

Reviewer 3: I made a small suggestion in the explantion of FinalizeBlock to address this.

Author: @lasarojc - the text should read something like this then?

Transactions are not guaranteed to be valid when they are delivered to an application, even if they were valid when they were proposed.

This can happen if the application state is used to determine transaction
validity. The application state may have changed between the initial execution of `CheckTx` and the transaction delivery in `FinalizeBlock` in a way that rendered the transaction no longer valid.

Note that `FinalizeBlock` **cannot** yet commit the Badger transaction we were building during the block execution.

 Other methods, such as `Query`, rely on a consistent view of the application's
 state, the application should only update its state by committing the Badger transactions when the full block has been delivered and the `Commit` method is invoked.

Reviewer 3: Looks great!

> is no deterministic way for it to make progress so the only safe option is to terminate.

Reviewer 3: ```suggestion
is no deterministic way for it to make progress so the only safe option is to terminate.
Once the application is restarted, the transactions in the block that failed execution will
be re-executed and should succeed if the Badger error was transient.
```

Note that `CheckTx` does not execute the transaction, it only verifies that the transaction could be executed. We do not know yet if the rest of the network has agreed to accept this transaction into a block.

Finally, make sure to add the `bytes` package to the `import` stanza at the top of `app.go`:

```go
import(
    "bytes"

    "github.com/dgraph-io/badger/v3"
    abcitypes "github.com/cometbft/cometbft/abci/types"
)
```

### 1.3.3 FinalizeBlock

`FinalizeBlock` is an ABCI method introduced in CometBFT v0.38.0. This replaces the functionality provided previously (pre-v0.38.0) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.

This method is responsible for executing the block and returning a response to the consensus engine.
The `ProcessProposal` method provides the transactions to the state machine before the block is finalized.
This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. 
Instead, the application (state machine) can choose to optimistically execute even before the block is finalized.

> Providing a single `FinalizeBlock` method to signal finalization of a block simplifies the ABCI interface and increases flexibility in the execution pipeline.

Reviewer 3: Good point. Another HOW-TO :)

            key, value := parts[0], parts[1]
            log.Printf("Adding key %s with value %s", key, value)

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }

            log.Printf("Successfully added key %s with value %s", key, value)

            // Marshal the ExecTxResult to the txs slice element
            txs[i] = &abcitypes.ExecTxResult{}
            if _, err := txs[i].MarshalToSizedBuffer(nil); err != nil {
                log.Panicf("Error marshalling tx result: %v", err)
            }
        }
    }

    return &abcitypes.ResponseFinalizeBlock{
    TxResults:        txs,
    AppHash:          []byte{},
    Events:           []abcitypes.Event{},
    ValidatorUpdates: abcitypes.ValidatorUpdates{},
    }, nil
}
```

The `Commit` method tells the application to make permanent the effects of
the application transactions.
Let's update the method to terminate the pending Badger transaction and
persist the resulting state:

```go
func (app KVStoreApplication) Commit(_ context.Context, commit *abcitypes.RequestCommit) (*abcitypes.ResponseCommit, error) {

>     if err := app.onGoingBlock.Commit(); err != nil {
>         return &abcitypes.ResponseCommit{}, err
>     }
>     return &abcitypes.ResponseCommit{}, nil

Reviewer 2: ```suggestion
    return &abcitypes.ResponseCommit{}, app.onGoingBlock.Commit()
```

Even if both are equivalent (CometBFT will panic anyway if `err` is not `nil`), the old version is easier to follow for a beginner.

Note that `CheckTx` does not execute the transaction, it only verifies that the transaction could be executed. We do not know yet if the rest of the network has agreed to accept this transaction into a block.

Finally, make sure to add the bytes package to the `import` stanza at the top of `app.go`:

```go
import(
"bytes"

"github.com/dgraph-io/badger/v3"
abcitypes "github.com/cometbft/cometbft/abci/types"
)
```

### 1.3.3 FinalizeBlock

> `FinalizeBlock` is a method introduced in CometBFT v0.38.0 which replaces the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock`.

Reviewer: ```suggestion
`FinalizeBlock` is an ABCI method introduced in CometBFT v0.38.0. This replaces the functionality provided previously (pre-v0.38.0) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.
```

> This method is responsible for executing the block and returning a response to the consensus engine. By implementing `ProcessProposal`, transactions are processed before the block execution phase and are already present in the state machine. This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. This simplifies the ABCI interface and increases flexibility in the execution pipeline.

Reviewer: I'm not sure if this is entirely accurate. Potential rephrase:

```suggestion
This method is responsible for executing the block and returning a response to the consensus engine. The `ProcessProposal` method provides the transactions to the state machine before the block is finalized. This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. Instead, the application (state machine) can choose to optimistically execute even before the block is finalized. Providing a single `FinalizeBlock` method to signal finalization of a block simplifies the ABCI interface and increases flexibility in the execution pipeline.
```

Note that `CheckTx` does not execute the transaction, it only verifies that the transaction could be executed. We do not know yet if the rest of the network has agreed to accept this transaction into a block.

Finally, make sure to add the bytes package to the `import` stanza at the top of `app.go`:

```go
import(
    "bytes"

    "github.com/dgraph-io/badger/v3"
    abcitypes "github.com/cometbft/cometbft/abci/types"
)
```

### 1.3.3 FinalizeBlock

> `FinalizeBlock` is an ABCI method introduced in CometBFT v0.38.0. This replaces the functionality provided previously (pre-v0.38.0) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.

Author: @lasarojc not sure if I understand your comment. The guide though introductory, is an "extension" of the previous 0.34 & 0.37 guide. 

Also, people will use 0.34 for some time before moving to 0.38. This might address their confusion about why `BeginBlock`, `DeliverTx`, and `EndBlock` are missing in 0.38. 

Or do you have a different concern?

> `FinalizeBlock` is an ABCI method introduced in CometBFT v0.38.0. This replaces the functionality provided previously (pre-v0.38.0) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.

Reviewer 3: My concern is that this adds unnecessary complexity for someone just starting on CometBFT.
You are right that someone coming from the an older version would benefit from this comment, but in this case, it be better to have simpler "Upgrading to ABCI2.0" guide. 
But if we don't have bandwidth for that, then a compromise, in my opinion, would be to have these references to previous versions as notes as in

> :note:
> `FinalizeBlock` is an ABCI method introduced in CometBFT v0.38.0. This replaces the functionality provided previously (pre-v0.38.0) by the combination of ABCI methods `BeginBlock`, `DeliverTx`, and `EndBlock`.

> This method is responsible for executing the block and returning a response to the consensus engine. The `ProcessProposal` method provides the transactions to the state machine before the block is finalized. This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. Instead, the application (state machine) can choose to optimistically execute even before the block is finalized. Providing a single `FinalizeBlock` method to signal finalization of a block simplifies the ABCI interface and increases flexibility in the execution pipeline.

Author: Thank you for adding periods.

> This method is responsible for executing the block and returning a response to the consensus engine. The `ProcessProposal` method provides the transactions to the state machine before the block is finalized. This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. Instead, the application (state machine) can choose to optimistically execute even before the block is finalized. Providing a single `FinalizeBlock` method to signal finalization of a block simplifies the ABCI interface and increases flexibility in the execution pipeline.

Author: Ah! Interesting. _Suggestion: Should we leave this line as is and link to some blog/article/guide in future?_ Or you have a stronger inclination towards removing this line ?

> The `FinalizeBlock` method executes the block, including any necessary transaction processing and state updates, and returns a `ResponseFinalizeBlock` object which contains any necessary information about the executed block.

Author: Fair point. Thanks for the suggestion!

```go
func (app *KVStoreApplication) FinalizeBlock(_ context.Context, req *abcitypes.RequestFinalizeBlock) (*abcitypes.ResponseFinalizeBlock, error) {
    var txs = make([]*abcitypes.ExecTxResult, len(req.Txs))

    app.onGoingBlock = app.db.NewTransaction(true)
    for i, tx := range req.Txs {
        if code := app.isValid(tx); code != 0 {
            log.Printf("Error in tx in if")
            txs[i] = &abcitypes.ExecTxResult{Code: code}
        } else {
            parts := bytes.SplitN(tx, []byte("="), 2)
            key, value := parts[0], parts[1]
            log.Printf("Adding key %s with value %s", key, value)

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }

            log.Printf("Successfully added key %s with value %s", key, value)

            // Marshal the ExecTxResult to the txs slice element
            txs[i] = &abcitypes.ExecTxResult{}

>             if _, err := txs[i].MarshalToSizedBuffer(nil); err != nil {

Reviewer 3: why this check here, but not in the case of returning an error?

    app.onGoingBlock = app.db.NewTransaction(true)
    for i, tx := range req.Txs {
        if code := app.isValid(tx); code != 0 {
            log.Printf("Error in tx in if")
            txs[i] = &abcitypes.ExecTxResult{Code: code}
        } else {
            parts := bytes.SplitN(tx, []byte("="), 2)
            key, value := parts[0], parts[1]
            log.Printf("Adding key %s with value %s", key, value)

            if err := app.onGoingBlock.Set(key, value); err != nil {
                log.Panicf("Error writing to database, unable to execute tx: %v", err)
            }

            log.Printf("Successfully added key %s with value %s", key, value)

            // Marshal the ExecTxResult to the txs slice element
            txs[i] = &abcitypes.ExecTxResult{}
            if _, err := txs[i].MarshalToSizedBuffer(nil); err != nil {
                log.Panicf("Error marshalling tx result: %v", err)
            }
        }
    }

    return &abcitypes.ResponseFinalizeBlock{
    TxResults:        txs,
    AppHash:          []byte{},
    Events:           []abcitypes.Event{},
    ValidatorUpdates: abcitypes.ValidatorUpdates{},
    }, nil
}
```

> Note that we check the validity of the transaction _again_ during `DeliverTx`.

Reviewer 3: I think this is important. It could be captured in a best practices document or left here.

>     return &abcitypes.ResponseCommit{}, err

Reviewer 3: ```suggestion
        return &abcitypes.ResponseCommit{}, err
```

This means that there is no need for the traditional ABCI methods of `BeginBlock`, `DeliverTx`, and `EndBlock` to handle the transaction execution and state updates. 
Providing a single `FinalizeBlock` method to signal the finalization of a block simplifies the ABCI interface and increases flexibility in the execution pipeline.

The `FinalizeBlock` method executes the block, including any necessary transaction processing and state updates, and returns a `ResponseFinalizeBlock` object which contains any necessary information about the executed block.

**Note:** `FinalizeBlock` only prepares the update to be made and does not change the state of the application. The change of state happens in a later stage i.e. is `commit` phase.

Note that, to implement these calls in our application we're going to make use of Badger's transaction mechanism. We will always refer to these as Badger transactions, not to confuse them with the transactions included in the blocks delivered by CometBFT, the _application transactions_.

First, let's create a new Badger transaction during `FinalizeBlock`. All application transactions in the current block will be executed within this Badger transaction.

> Next, let's modify `FinalizeBlock` to add the `key` and `value` to the database transaction every time our application receives a new application transaction through `RequestFinalizeBlock`.

Reviewer 2: ```suggestion
Next, let's modify `FinalizeBlock` to add the `key` and `value` to the database transaction every time our application processes a new application transaction from the list received through `RequestFinalizeBlock`.
```

            log.Printf("Successfully added key %s with value %s", key, value)

            txs[i] = &abcitypes.ExecTxResult{}
        }
    }

    return &abcitypes.ResponseFinalizeBlock{
      TxResults:        txs,
    }, nil
}
```

Transactions are not guaranteed to be valid when they are delivered to an application, even if they were valid when they were proposed.

This can happen if the application state is used to determine transaction validity. 

> The application state may have changed between the initial execution of CheckTx and the transaction delivery in FinalizeBlock in a way that rendered the transaction no longer valid.

Reviewer 2: ```suggestion
The application state may have changed between the initial execution of `CheckTx` and the transaction delivery in FinalizeBlock in a way that rendered the transaction no longer valid.
```

> **Note** that FinalizeBlock cannot yet commit the Badger transaction we were building during the block execution.

Reviewer 2: ```suggestion
**Note** that `FinalizeBlock` cannot yet commit the Badger transaction we were building during the block execution.
```

> Other methods, such as Query, rely on a consistent view of the application's state, the application should only update its state by committing the Badger transactions when the full block has been delivered and the Commit method is invoked.

Reviewer 2: ```suggestion
Other methods, such as `Query`, rely on a consistent view of the application's state, the application should only update its state by committing the Badger transactions when the full block has been delivered and the Commit method is invoked.
```

Reviewer 2: > by committing the Badger transactions

Is there one or several Badger transactions?

to modify the group by invoking `PrepareProposal`.

The application is free to modify the group before returning from the call, as long as the resulting set
does not use more bytes than `RequestPrepareProposal.max_tx_bytes`
For example, the application may reorder, add, or even remove transactions from the group to improve the
execution of the block once accepted.
In the following code, the application simply returns the unmodified group of transactions:

```go
func (app *KVStoreApplication) PrepareProposal(_ context.Context, proposal *abcitypes.RequestPrepareProposal) (*abcitypes.ResponsePrepareProposal, error) {

>     return &abcitypes.ResponsePrepareProposal{Txs: proposal.Txs}, nil

Reviewer 2: There is a problem here, which is also present before this PR.
The spec requires the App to make sure the size of `Txs` does not exceed `RequestPrepareProposal.MaxTxByte`. So a correct implementation should go through the received transactions and exclude those that don't fit.
For an example on how we do that, check [this](https://github.com/cometbft/cometbft/blob/v0.37.x/test/e2e/app/app.go#L273-L290)

Reviewer 2: You may want to tackle this comment in a different PR, so it can be backported to both `v0.38.x` and `v0.37.x`

In the following code, the application simply returns the unmodified group of transactions:

```go

> func (app *KVStoreApplication) PrepareProposal(_ context.Context, proposal *abcitypes.RequestPrepareProposal) (*abcitypes.ResponsePrepareProposal, error) {

Reviewer 3: In line 446, above, where you see "does not use more bytes than `RequestPrepareProposal.max_tx_bytes'" you should have "does not use more bytes than `RequestPrepareProposal.max_tx_bytes`" with a backtick instead of single quotes at the end.

Author: Ah! Good catch. Thanks.

    db, err := badger.Open(badger.DefaultOptions(dbPath))
    
    if err != nil {
        log.Fatalf("Opening database: %v", err)
    }
    defer func() {
    if err := db.Close(); err != nil {
        log.Printf("Closing database: %v", err)
    }
}()

app := NewKVStoreApplication(db)

pv := privval.LoadFilePV(
    config.PrivValidatorKeyFile(),
    config.PrivValidatorStateFile(),
)

nodeKey, err := p2p.LoadNodeKey(config.NodeKeyFile())
if err != nil {
    log.Fatalf("failed to load node's key: %v", err)
}

logger := cmtlog.NewTMLogger(cmtlog.NewSyncWriter(os.Stdout))
logger, err = cmtflags.ParseLogLevel(config.LogLevel, logger, cfg.DefaultLogLevel)

if err != nil {
    log.Fatalf("failed to parse log level: %v", err)
}
node, err := nm.NewNode(
    config,
    pv,
    nodeKey,
    proxy.NewLocalClientCreator(app),
    nm.DefaultGenesisDocProviderFunc(config),
    nm.DefaultDBProvider,
    nm.DefaultMetricsProvider(config.Instrumentation),
logger)

if err != nil {
    log.Fatalf("Creating node: %v", err)
}
node.Start()
defer func() {
    node.Stop()
    node.Wait()
}()

c := make(chan os.Signal, 1)
signal.Notify(c, os.Interrupt, syscall.SIGTERM)

> <-c

Author: There are no changes in main from previous code.
